File: CSharpCodeBase/core/gameloop.lua
local Class = require 'utils.hump.class'

local GameLoop = Class({init = function (self, gameObject)
      self.items = {}
      self.affectQueue = {}
end})

function GameLoop:FixedUpdate()
  for k, v in pairs(self.items) do
    if v.enabled then
      v:FixedUpdate()
    end
  end
  self:ProcessAffectQueue()
end

function GameLoop:LateUpdate()
  for k, v in pairs(self.items) do
    if v.enabled then
      v:LateUpdate()
    end
  end
  self:ProcessAffectQueue()
end

function GameLoop:Update()
  --local updateCount = 0
  for k, v in pairs(self.items) do
    if v.enabled then
      --updateCount = updateCount + 1
      v:Update()
    end
  end    
  self:ProcessAffectQueue()
  --print(updateCount)
end

function GameLoop:Add(item)
  table.insert(self.affectQueue, {item, 1})
  --self.affectQueue[#self.affectQueue + 1] = {item, 1}
end

function GameLoop:AddForce(item)
  local idx = Tables.Find(self.items, item)
  if idx < 0 then
    table.insert(self.items, item)
    --self.items[#self.items + 1] = item
  end
end

function GameLoop:Remove(item)
  item.enabled = false
  self.affectQueue[#self.affectQueue + 1] = {item, 0}
end

function GameLoop:ProcessAffectQueue()
  if #self.affectQueue > 0 then
    for k, v in pairs(self.affectQueue) do
      if v[2] == 0 then
        self:RemoveForce(v[1])
      else
        self:AddForce(v[1])
      end
    end
    self.affectQueue = {}
  end
end

function GameLoop:RemoveForce(item)
  local idx = Tables.Find(self.items, item)
  if idx > -1 then
    table.remove(self.items, idx)
  end
end

return GameLoop

File: CSharpCodeBase/core/gameloop.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class gameloop {
 public void init (self, gameObject);
       this.items = {}
       this.affectQueue = {}
 }})
 public void FixedUpdate(){
   for(k, v in pairs(this.items) ){
     if(v.enabled  ){ 
       v:FixedUpdate();
     }
   }
   self:ProcessAffectQueue();
 }
 public void LateUpdate(){
   for(k, v in pairs(this.items) ){
     if(v.enabled  ){ 
       v:LateUpdate();
     }
   }
   self:ProcessAffectQueue();
 }
 public void Update(){
   //local updateCount = 0
   for(k, v in pairs(this.items) ){
     if(v.enabled  ){ 
       //updateCount = updateCount + 1
       v:Update();
     }
   }    
   self:ProcessAffectQueue();
   //print(updateCount)
 }
 public void Add(item){
   table.insert(this.affectQueue, {item, 1})
   //this.affectQueue[#this.affectQueue + 1] = {item, 1}
 }
 public void AddForce(item){
   var idx = Tables.Find(this.items, item);
   if(idx < 0  ){ 
     table.insert(this.items, item);
     //this.items[#this.items + 1] = item
   }
 }
 public void Remove(item){
   item.enabled = false;
   this.affectQueue[#this.affectQueue + 1] = {item, 0}
 }
 public void ProcessAffectQueue(){
   if(#this.affectQueue > 0  ){ 
     for(k, v in pairs(this.affectQueue) ){
       if(v[2] == 0  ){ 
         self:RemoveForce(v[1]);
       }else{
         self:AddForce(v[1]);
       }
     }
     this.affectQueue = {}
   }
 }
 public void RemoveForce(item){
   var idx = Tables.Find(this.items, item);
   if(idx > -1  ){ 
     table.remove(this.items, idx);
   }
 }
 }}